### PR TITLE
Update to 0.8.11, add network-agnostic deployment script, add hardhat support, add package.json and dependency control

### PR DESCRIPTION
**Pull request by bghughes**
I believe this PR brings a number of improvements to the protocol, namely adding Hardhat and a network-agnostic deployment script that is extensible to any network. Please note some added scripts:
- `npm run deployOptimismKovan`: deploys from the `deploy` folder the Arrow ERC20 to selected hardhat network.
- `npm run verifyOptmismKovan`: verifies network deployments automatically on etherscan. Make sure to add "-- CONTRUCTORARG1, 2, etc." after the command in the terminal so the plugin knows constructor params.

Please note, Brownie CI is behaving oddly after I updated to Solidity 0.8.11 and started driving imports off of OpenZeppelin's published node packages (best practice AFAIK). I in no way wanted to break this, brining this up here for @BrassLion 's reference.

If you want @thomasgarrison, I can spin up an issue and try to port the testing suite to hardhat and add continuos integration on that front/real-time testing.

Please find the example Arrow contract on Optimism Kovan that can be deployed with the above scripts and custom parameters (edit deploy script) [here](https://kovan-optimistic.etherscan.io/address/0xABb4683397a0342893157aCDDC8d91C90473217F#code).

### CODE REVIEW

**Comment by bghughes** [comment]
Btw, I don't think there will be an issue w/ ProxyAdmin/ContractAdmin collision if we do it right. On the `TransparentUpgradeableProxy` standard this is solved using a dedicated ProxyAdmin contract and perhaps on this new UUPS implementation that OZ is using in 0.4 it might not be needed at all...

https://docs.openzeppelin.com/contracts/4.x/api/proxy#transparent-vs-uups

Please allow me to do some more research here.

**Comment by bghughes** [comment]
Closes #8

File: contracts/ArrowVestingBase.sol
Comment (BrassLion) [inline contracts/ArrowVestingBase.sol:5]: This change is breaking the existing CI tests, because brownie can no longer find the correct import path. Can we revert this? If not, we'll need to change the brownie import logic to cope with the change.

File: contracts/ArrowVestingFactory.sol
Comment (BrassLion) [inline contracts/ArrowVestingFactory.sol]:
```suggestion
    event NewVestingAgreement(
```
Events objects should probably be capitalised to denote their type and fit with convention.

File: deploy/deployArrowTokenHot.js
Comment (BrassLion) [inline deploy/deployArrowTokenHot.js:17]: Initial token supply needs to be decided and agreed on before deployment on mainnet. We should probably make an issue to track and address this.

File: deploy/deployArrowTokenHot.js
Comment (BrassLion) [inline deploy/deployArrowTokenHot.js:53]: Is there any way to easily add tests to verify the deployment will succeed?

**Review by BrassLion — DISMISSED** [review]
Generally LGTM, with my limited knowledge of Hardhat. I've added a couple of minor things that could be addressed. CI should pass before we merge this.

FYI, I have a competing deployment PR [here](brasslion/feature/deployment-script), so we should decide on the best deployment route and use one moving forward.

File: contracts/ArrowVestingBase.sol
Comment (BrassLion) [reply contracts/ArrowVestingBase.sol:5]: Brownie imports from Github releases, so we should probably just update the path substitution Brownie does to match the NPM package. You can find it in `brownie-config.yaml`.

File: contracts/ArrowVestingFactory.sol
Comment (GoodluckH) [inline contracts/ArrowVestingFactory.sol]: If `ArrowVestingBase` is upgradeable, can we really use `immutable` here? 

https://docs.openzeppelin.com/upgrades-plugins/1.x/faq#why-cant-i-use-immutable-variables

File: deploy/deployArrowTokenHot.js
Comment (GoodluckH) [reply deploy/deployArrowTokenHot.js:17]: Just opened up an issue ticket here: #11

File: contracts/ArrowVestingFactory.sol
Comment (BrassLion) [reply contracts/ArrowVestingFactory.sol]: I think this is fine, because `ArrowVestingFactory` is not an upgradable contract, so it can use immutable values without issue.

File: deploy/deployArrowTokenHot.js
Comment (bghughes) [reply deploy/deployArrowTokenHot.js:17]: Agreed - that is an adjustable plug for now. Resolving, thanks @GoodluckH

File: contracts/ArrowVestingBase.sol
Comment (bghughes) [reply contracts/ArrowVestingBase.sol:5]: Is this more sound than following official npm releases? I am trying to replicate this in hardhat overrides and now thinking to myself if npm is the better path... allows for better composability IMO

File: contracts/ArrowVestingFactory.sol
Comment (bghughes) [reply contracts/ArrowVestingFactory.sol]: Agreed @GoodluckH - should not be immutable

File: deploy/deployArrowTokenHot.js
Comment (bghughes) [reply deploy/deployArrowTokenHot.js:53]: Yes there is, I can do this

**Comment by bghughes** [comment]
> Generally LGTM, with my limited knowledge of Hardhat. I've added a couple of minor things that could be addressed. CI should pass before we merge this.
> 
> FYI, I have a competing deployment PR [here](brasslion/feature/deployment-script), so we should decide on the best deployment route and use one moving forward.

Thanks for your input! Very helpful. I just resolved a few of the items and am leaving the open issues unresolved. Getting this testing suite in Mocha and moving the CI and testing to Hardhat may be the next step - what do you think about that?

Once the CI is resolved and I write the verification check you asked for, this should be good to merge in :)

**Comment by GoodluckH** [comment]
> > Generally LGTM, with my limited knowledge of Hardhat. I've added a couple of minor things that could be addressed. CI should pass before we merge this.
> > FYI, I have a competing deployment PR [here](brasslion/feature/deployment-script), so we should decide on the best deployment route and use one moving forward.
> 
> Thanks for your input! Very helpful. I just resolved a few of the items and am leaving the open issues unresolved. Getting this testing suite in Mocha and moving the CI and testing to Hardhat may be the next step - what do you think about that?
> 
> Once the CI is resolved and I write the verification check you asked for, this should be good to merge in :)

Just wanted to make sure that we don't do duplicated works. I've already wrote a chai test suite (see  #12 [comment](https://github.com/Arrow-air/Arrow-Contracts/pull/12#issuecomment-1047320665)) and have created a GitHub Action for running `npx hardhat test`  whenever a push or pull request is made.

File: contracts/ArrowVestingFactory.sol
Comment (BrassLion) [inline contracts/ArrowVestingFactory.sol:13]: What does changing this to `public` achieve? I believe that reading `immutable` variables are cheaper from a gas POV than `public`.

File: contracts/ArrowVestingBase.sol
Comment (BrassLion) [reply contracts/ArrowVestingBase.sol:5]: I don't see the difference between importing the official npm releases or the offical GitHub releases, since they're both ultimately controlled by OpenZeppelin.

It seems like the prevailing wind is to entirely replace Brownie with Hardhat though, so it makes sense to use whatever works best with Hardhat's configuration and go from there. I'm happy with you just going with the npm import and not bend over backwards to support both frameworks, as long as the test coverage remains the same.

**Comment by BrassLion** [comment]
@bghughes Any ways I can help out on moving this along?

File: contracts/ArrowVestingFactory.sol
Comment (bghughes) [reply contracts/ArrowVestingFactory.sol:13]: immutable means we cannot change it, thus making the implementation un-upgradeable. Must be public in my opinion for the market to see it and for us to upgrade

File: deploy/deployArrowTokenHot.js
Comment (bghughes) [reply deploy/deployArrowTokenHot.js:53]: I made an open ticket for this in #14

**Pull request merged**
